### DISCUSSION TITLE
Count using `with_entities` returns wrong count when there is no filter

### DISCUSSION THREAD

**Discussion by nameer** [sqlalchemy/sqlalchemy#8788] [Usage Questions] [answered]
### Describe the bug

A wrong count is returned when the count is queried using `with_entitites`. The expected count is returned if the `count` method is used on the query. An expected result is also returned when a filter or column name is specified

### To Reproduce

```python
count_using_with_entities = (
    session.query(models.User).with_entities(func.count()).scalar()
)
count_using_count = session.query(models.User).count()
print(f"{count_using_with_entities=}")  # wrong count
print(f"{count_using_count=}")  # correct count
print()

# If I specify a column name, the count is correct.
count_using_with_entities_col_name = (
    session.query(models.User).with_entities(func.count(models.User.id)).scalar()
)
print(f"{count_using_with_entities_col_name=}")
print()

# If I add a filter, the result is correct.
filtered_count_using_with_entities = (
    session.query(models.User)
    .filter(models.User.email.ilike("%user.com"))
    .with_entities(func.count())
    .scalar()
)
filtered_count_using_count = (
    session.query(models.User).filter(models.User.email.ilike("%user.com")).count()
)
print(f"{filtered_count_using_with_entities=}")  # correct count
print(f"{filtered_count_using_count=}")  # correct count
```

### Error

Output:

```
count_using_with_entities=1
count_using_count=3

count_using_with_entities_col_name=3

filtered_count_using_with_entities=2
filtered_count_using_count=2
```

Generated SQL statements:
- count_using_with_entities:
  ```SQL
  SELECT count(*) AS count_1
  ```
  I see no table name is specified here, even though I have specified the table in the SQLAlchemy during the query.
- count_using_count:
  ```SQL
  SELECT count(*) AS count_1 
  FROM (SELECT "user".id AS user_id, "user".first_name AS user_first_name, "user".last_name AS user_last_name, "user".email AS user_email, "user".is_active AS user_is_active
  FROM "user") AS anon_1
  ```
- count_using_with_entities_col_name:
  ```SQL
  SELECT count("user".id) AS count_1 
  FROM "user"
  ```
- filtered_count_using_with_entities:
  ```SQL
  SELECT count(*) AS count_1 
  FROM "user" 
  WHERE "user".email ILIKE %(email_1)s
  ```
- filtered_count_using_count:
  ```SQL
  SELECT count(*) AS count_1 
  FROM (SELECT "user".id AS user_id, "user".first_name AS user_first_name, "user".last_name AS user_last_name, "user".email AS user_email, "user".is_active AS user_is_active 
  FROM "user" 
  WHERE "user".email ILIKE %(email_1)s) AS anon_1
  ```

### Versions

- OS: Ubuntu 22.04.1 LTS
- Docker OS: Debian GNU/Linux 11 (bullseye)
- Python: 3.9.7
- SQLAlchemy: 1.4.42
- Database: Postgres
- DBAPI: psycopg2

### Additional context

I cannot always use a filter since the API may be called without a filter. Also, since the class I'm writing is a generic one, I cannot specify a column name either.

**Comment by CaselIT** [comment]
Hi,

I believe this is the expected behaviour:
I would expect `session.query(models.User).with_entities(func.count()).scalar()` to generate `SELECT count(*) AS count_1` since `with_entities` replaces the selected entities.

correct me if I'm wrong @zzzeek

**Reply by zzzeek** [reply]
that's correct

**Reply by nameer** [reply]
Hmm... I would expect `session.query(models.User).with_entities(func.count())` to select from the table `User` (`SELECT count(*) AS count_1 FROM "user"`) since I have already mentioned the table name

**Reply by CaselIT** [reply]
The would not match the documentation, since it says "Return a new Query replacing the SELECT list with the given entities."

Doing `session.query(models.User).with_entities(func.count())` is equivalent to doing `session.query(func.count())`

**Reply by nameer** [reply]
That's exactly what my point is. If we wanted `SELECT count(*) AS count_1`, we could have used `session.query(func.count())`. But when `session.query(models.User).with_entities(func.count())` is queried, `models.User` is neglected. I think it is a bit confusing.

**Reply by zzzeek** [reply]
this is why designing a system by adding methods as you think of them eventually becomes a house of cards

**Comment by zzzeek** [comment, marked as answer]
the entire system of `query.count()` is one of the major pillars of why `Query` is legacy.  if you are looking for consistent SQL SELECT behavior, I recommend you migrate to [2.0 style queries](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-orm-usage).  for the behaviors above, it's also essential that you set [echo=True](https://docs.sqlalchemy.org/en/14/core/engines.html#more-on-the-echo-flag) so that you can see what queries are being emitted.  for `session.query(models.User).with_entities(func.count()).scalar()` for example, this would reveal that `with_entities()` is replacing the FROM clause of the query, which you can set by calling `s.query(User).with_entities(func.count()).select_from(User).scalar()`.

**Reply by nameer** [reply]
Adding `select_from` works like a charm ✋🏽 Though I expected `session.query(models.User).with_entities(func.count())` to count from the `User` table since I have already mentioned the table name. Adding `select_from` seems unnecessary to me.

And, I have not checked out 2.0, will definitely do that 😃 👍🏽

**Comment by qihboy** [comment]
I've patched this issue in my projects like this:

cause `select count(*) from (select * from table where...) ` not so good at sql query as `select count(*) from table where`

query = ....some....complex...or...simple...query....

```python
statement = query.statement
from_tables = statement.froms
where_clause = statement.whereclause
count_query = self.statement.with_only_columns([func.count()]).order_by(None)
if not where_clause and len(from_tables) == 1:
    count_query = count_query.select_from(from_tables[0])


total = db.session.execute(count_query).scalar()
```

**Reply by CaselIT** [reply]
All that is not needed. You can use this parameter of with_only_columns https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.with_only_columns.params.maintain_column_froms

**Reply by qihboy** [reply]
Yes, yours is the explicit way, I didn't notice that param before.

```python
count_query = query.statement.with_only_columns([func.count()], maintain_column_froms=True).order_by(None)
```
